### PR TITLE
Fix GIL related error in pybind11 interface.

### PR DESCRIPTION
**Pull request by thorstenhater**
Fixes an issue in the python bindings, that appeared when upgrading to pybind11 v2.6.1, 
see https://github.com/arbor-sim/arbor/issues/1271.

This is triggered by the upgrade, but is actually incorrect in any case. Here is why:

- (PY) Simulation::init releases the GIL
- (CXX) Simulation calls through to the Recipe
- (PY) A python class deriving from (CXX) Recipe returns a python object.
- This (PY) object is then destroyed in python land, which requires the GIL.

Pybind11 only guarantees acquisition of the GIL at edges C++ -> Python.
So, we need an extra GIL acquisition here. Our usecase is actually an example
here
https://pybind11.readthedocs.io/en/stable/advanced/misc.html?highlight=gil#global-interpreter-lock-gil.

Note: We use `OVERLOAD_PURE` here, which acquires the GIL at the edge Python -> C++
for the duration of the wrapped call, *but* that does not cover the destruction of the return
value, thus triggering the bug.

I tested the fix with multiple threads on my machine, but it's quite hard to test.

T

Update:
In addition -- like mentioned in #1271 -- event generators have been fixed. None of the other
methods in recipe produce this issues. Only simulation releases the GIL, no other trampolines 
are used. This should eliminate all of the issues.

### CODE REVIEW

File: python/recipe.cpp [hunk 1]
```diff
@@ -21,22 +23,53 @@
 
 namespace pyarb {
 
+// Convert a cell description inside a Python object to a cell description in a
+// unique_any, as required by the recipe interface.
+// This helper is only to be called while holding the GIL We require this guard
```
Comment (bcumming) [inline python/recipe.cpp:28]: Full stop in `GIL. We`

File: python/recipe.cpp [hunk 2]
```diff
@@ -21,22 +23,53 @@
...
+// across the lifetime of the description object `d`, since this function can be
```
Comment (bcumming) [inline python/recipe.cpp:29]: What is `d`? Do you mean `o`?

File: python/recipe.cpp [hunk 3]
```diff
@@ -58,9 +91,12 @@ std::vector<arb::event_generator> convert_gen(std::vector<pybind11::object> pyge
 }
 
 std::vector<arb::event_generator> py_recipe_shim::event_generators(arb::cell_gid_type gid) const {
-    return try_catch_pyexception(
-                [&](){ return convert_gen(impl_->event_generators(gid), gid); },
-                "Python error already thrown");
+    return try_catch_pyexception([&](){
+        pybind11::gil_scoped_acquire guard;
+        auto result = convert_gen(impl_->event_generators(gid), gid);
```
Comment (bcumming) [inline python/recipe.cpp:96]: Any reason why we don't just `return convert_gen(...);`?

**Review by bcumming — CHANGES_REQUESTED** [review]
Great work tracking down this nasty latent bug.

There are just a couple of nit-picks.

File: python/recipe.cpp (hunk 3 above)
Comment (thorstenhater) [reply python/recipe.cpp:96]: Good point.

File: python/recipe.cpp (hunk 2 above)
Comment (thorstenhater) [reply python/recipe.cpp:29]: :+1:

File: python/recipe.cpp (hunk 1 above)
Comment (thorstenhater) [reply python/recipe.cpp:28]: Yes.

**Review by bcumming — APPROVED** [review]

**Pull request merged**